Added quality declaration draft (#142)
ros/class_loader: QUALITY_DECLARATION.md

This document is a declaration of software quality for the `class_loader` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).

# `class_loader` Quality Declaration

> The package `class_loader` claims to be in the **Quality Level 1** category.

Reviewer: I'm not sure this quite makes it yet to QL 1 given some missing chunks below.

Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 1 in REP-2004](https://www.ros.org/reps/rep-2004.html).

## Version Policy [1]

### Version Scheme [1.i]

> `class_loader` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning), and is at or above a stable version. **The current version is 1.4.0**.

Reviewer: > **The current version is 1.4.0**.

Since versions change quite a bit, I would just change this to suggest where to find the current version.

Also it seems like this package just named itself 1.4 because ROS 1 uses 0.4. Does 1.x actually imply API stability?

Author: Not sure about this. I assumed that the current version in the package.xml is the right one. Maybe for ROS 2 we need to restart from "zero" the version number.

```
  <version>1.4.0</version>
```

### Version Stability [1.ii]

### Public API Declaration [1.iii]

All symbols in the installed headers are considered part of the public API.

All installed headers are in the `include` directory of the package, headers in any other folders are not installed and considered private.

### API Stability Within a Released ROS Distribution [1.iv]/[1.vi]

`class_loader` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Within a Released ROS Distribution [1.v]/[1.vi]

`class_loader` will maintain ABI stability within a ROS distribution.

## Change Control Process [2]

`class_loader` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change_control_process).

### Change Requests [2.i]

All changes occur through a pull request

### Contributor Origin [2.ii]

> Must have confirmation of contributor origin (e.g. DCO, CLA, etc.).

Reviewer: This should be rephrased to state what DCO process it follows, not just repeat the info in rep-2004.

### Peer Review Policy [2.iii]

> All pull request have two peer reviews

Reviewer: It looks like this repo only requires 1 peer review currently. Are there plans to change it?

### Continuous Integration [2.iv]

> all pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Reviewer: ```suggestion
All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
```

### Documentation Policy [2.v]

> All pull request must resolve related documentation changes before merging

Reviewer: ```suggestion
All pull requests must resolve related documentation changes before merging
```

## Documentation [3]

### Feature Documentation [3.i]

> `class_loader` has a [feature list](http://wiki.ros.org/class_loader) and each item in the list links to the corresponding feature documentation.

Reviewer: This is the ROS 1 feature list, we need one for ROS 2. Or at the very least, state something to the effect of "There is no current ROS 2 feature list, but the feature list for ROS 1 can be found at..."

There is documentation for all of the features, and new features require documentation before being added.

### Public API Documentation [3.ii]

> `class_loader` has embedded API documentation and it is generated using doxygen and is [hosted](http://wiki.ros.org/class_loader) along side the feature documentation.

Reviewer: I don't think this should link to ros1 docs, and ROS 2 API docs aren't currently being generated for class_loader

https://github.com/ros2/docs.ros2.org/blob/a5ed03490741da161598f3b639d47eb8f361984e/Makefile#L4-L14

There is an automated test which runs a linter (ament_copyright) that ensures each file has a license statement.

### Copyright Statement [3.iv]

The copyright holders each provide a statement of copyright in each source code file in `class_loader`.

There is an automated test which runs a linter (ament_copyright) that ensures each file has at least one copyright statement.

## Testing [4]

### Feature Testing [4.i]

Each feature in `class_loader` has corresponding tests which simulate typical usage, and they are located in the `test` directory.
New features are required to have tests before being added.

### Public API Testing [4.ii]

Each part of the public API have tests, and new additions or changes to the public API require tests before being added.
The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.

### Coverage [4.iii]

`class_loader` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#coverage), and opts to use branch coverage instead of line coverage.

This includes:

- tracking and reporting branch coverage statistics

> - achieving and maintaining branch coverage at or above 95%

Reviewer: I think the current coverage is 90-ish % yah?

- no lines are manually skipped in coverage calculations

Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.

Current coverage statistics can be viewed here:

> ![](https://codecov.io/gh/ahcorde/class_loader/branch/ahcorde_code_coverage/graphs/sunburst.svg)

Reviewer: This image isn't too helpful. Maybe you could run a ci_linux-coverage job and link that instead.

### Performance [4.iv]

> `class_loader` follows the recommendations for performance testing of C++ code in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#performance-testing), and opts to do performance analysis on each release rather than each change.

Reviewer: Where are the performance tests? Just state the things that are true currently and provide links with supporting evidence.

### Linters and Static Analysis [4.v]

`class_loader` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).

## Dependencies [5]

> ### Direct Runtime ROS Dependencies [5.i]

Reviewer: This package does depend on rcpputils

### Optional Direct Runtime ROS Dependencies [5.ii]

### Direct Runtime non-ROS Dependency [5.iii]

`class_loader` has run-time and build-time dependencies that need to be considered for this declaration.

>  - console_bridge_vendor

Reviewer: console_bridge_vendor is technically a ROS dep yah?

 - libconsole-bridge-dev

It has one "buildtool" dependency, which do not affect the resulting quality of the package, because they do not contribute to the public library API.
It also has several test dependencies, which do not affect the resulting quality of the package, because they are only used to build and run the test code.

## Platform Support [6]

> `class_loader` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.

Reviewer: Add links to latest CI results. and remove TODO. I don't think we need to say anything about non-tier 1 unless we have ci tests for them.

This document is a declaration of software quality for the `class_loader` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).

# `class_loader` Quality Declaration

The package `class_loader` claims to be in the **Quality Level 2** category.

Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 2 in REP-2004](https://www.ros.org/reps/rep-2004.html).

## Version Policy [1]

### Version Scheme [1.i]

`class_loader` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning), and is at or above a stable version.

### Version Stability [1.ii]

> The current version is 1.4.0.

Reviewer: Like I mentioned, I don't think we want to have the specific version here unless we can automate it as this will get behind almost immediately when they bump versions and release it. I would just say something like, "It is at a stable version and currently above 1.0.0. See the [package.xml](package.xml) for it's current version."

### API Stability Within a Released ROS Distribution [1.iv]/[1.vi]

`class_loader` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Within a Released ROS Distribution [1.v]/[1.vi]

`class_loader` will maintain ABI stability within a ROS distribution.

## Change Control Process [2]

`class_loader` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change_control_process).

### Change Requests [2.i]

This package requires that all changes occurr through a pull request.

### Contributor Origin [2.ii]

This package has a confirmation of contributor origin policy, which can be found in [CONTRIBUTING](./CONTRIBUTING.md).

### Peer Review Policy [2.iii]

Following the recommended guidelines for ROS Core packages, all pull requests must have at least 1 peer review.

### Continuous Integration [2.iv]

All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Currently nightly results can be seen here:

> * [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)

Reviewer: Point to the ubuntu one instead (https://ci.ros2.org/job/nightly_linux_release/)

> * [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/class_loader/)

Reviewer: This says arm64, but it points to nightly_linux_release, switch the link to arm64.

* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/class_loader/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/class_loader/)

### Documentation Policy [2.v]

All pull requests must resolve related documentation changes before merging

## Documentation [3]

### Feature Documentation [3.i]

There is no current ROS 2 feature list, but the feature list for ROS 1 can be found [here](http://wiki.ros.org/class_loader) and each item in the list links to the corresponding feature documentation.

> There is documentation for all of the features, and new features require documentation before being added.

Reviewer: > There is documentation for all of the features, and new features require documentation before being added.

This seems misleading if there is not yet a ROS 2 features list, and it would be impossible to enforce the second part. Maybe just remove this sentence for now.

### License [3.iii]

The license for `class_loader` is BSD-3-Clause, and a summary is in each source file, the type is declared in the [package.xml](./package.xml) manifest file, and a full copy of the license is in the [LICENSE](./LICENSE) file.

There is an automated test which runs a linter (ament_copyright) that ensures each file has a license statement.

### Copyright Statement [3.iv]

The copyright holders each provide a statement of copyright in each source code file in `class_loader`.

There is an automated test which runs a linter (ament_copyright) that ensures each file has at least one copyright statement.

Most recent test results can be found [here](http://build.ros2.org/view/Epr/job/Epr__class_loader__ubuntu_bionic_amd64/lastBuild/testReport/class_loader/copyright_check_tests_only/)

## Testing [4]

### Feature Testing [4.i]

Each feature in `class_loader` has corresponding tests which simulate typical usage, and they are located in the `test` directory.

Currently nightly results can be seen here:

> * [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)

Reviewer: Same comment as above

* [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/class_loader/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/class_loader/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/class_loader/)

### Public API Testing [4.ii]

Each part of the public API have tests, and new additions or changes to the public API require tests before being added.
The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.

### Coverage [4.iii]

`class_loader` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#coverage), and opts to use branch coverage instead of line coverage.

This includes:

- tracking and reporting line coverage statistics

> - achieving and maintaining a reasonable branch line coverage (90-100%)

Reviewer: Since 90% doesn't quite meet the ROS 2 core guidelines, I would state this as something more like, "This package does not yet meet the 95% coverage guideline, but it is currently above XY%

- no lines are manually skipped in coverage calculations

Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.

> Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/lastBuild/cobertura/).

Reviewer: Point to the specific build number (85 in this case) so that the link remains durable. Also, since this was a build of everything, I would also refine the link to point to the class_loader specific stuff so it's easier to verify the results. Also these two results suggest the coverage is more in the 80s. Is there something I'm missing?

https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/src_ros_class_loader_include_class_loader/
https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/src_ros_class_loader_src/

Author: I use last build to know the current status. But I'm okey with a static value.

Fixed links

### Performance [4.iv]

`class_loader` does not currently have performance tests.

### Linters and Static Analysis [4.v]

`class_loader` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).

## Dependencies [5]

`class_loader` has run-time and build-time dependencies that need to be considered for this declaration.

### Direct Runtime ROS Dependencies [5.i]/[5.ii]

- console_bridge_vendor

> - rcpputils

Reviewer 2: `rcpputils` is [claiming level 4](https://github.com/ros2/rcpputils/pull/47/files#diff-92440392597dfc3b7adb67bc4ed481b3R5), so this package can't be at a lower level than its direct runtime dependencies.

That's why it's important we write the declaration for the lower level packages first.

### Direct Runtime non-ROS Dependency [5.iii]

 - libconsole-bridge-dev

It has one "buildtool" dependency, which do not affect the resulting quality of the package, because they do not contribute to the public library API.
It also has several test dependencies, which do not affect the resulting quality of the package, because they are only used to build and run the test code.

## Platform Support [6]

`class_loader` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.

Currently nightly results can be seen here:

> * [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)

Reviewer: Same comment as above

### Version Stability [1.ii]

The current version is 1.4.0.

### Public API Declaration [1.iii]

All symbols in the installed headers are considered part of the public API.

All installed headers are in the `include` directory of the package, headers in any other folders are not installed and considered private.

### API Stability Within a Released ROS Distribution [1.iv]/[1.vi]

`class_loader` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Within a Released ROS Distribution [1.v]/[1.vi]

`class_loader` will maintain ABI stability within a ROS distribution.

## Change Control Process [2]

`class_loader` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change_control_process).

### Change Requests [2.i]

This package requires that all changes occurr through a pull request.

### Contributor Origin [2.ii]

This package has a confirmation of contributor origin policy, which can be found in [CONTRIBUTING](./CONTRIBUTING.md).

### Peer Review Policy [2.iii]

> Following the recommended guidelines for ROS Core packages, all pull request have at least 1 peer review.

Reviewer: ```suggestion
Following the recommended guidelines for ROS Core packages, all pull requests must have at least 1 peer review.
```

### Continuous Integration [2.iv]

> All pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Reviewer: ```suggestion
All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
```

### Coverage [4.iii]

`class_loader` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#coverage), and opts to use branch coverage instead of line coverage.

This includes:

- tracking and reporting line coverage statistics
- achieving and maintaining a reasonable branch line coverage (90-100%)
- no lines are manually skipped in coverage calculations

Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.

Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/lastBuild/cobertura/).

### Performance [4.iv]

`class_loader` does not currently have performance tests.

### Linters and Static Analysis [4.v]

`class_loader` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).

## Dependencies [5]

> ### Direct Runtime ROS Dependencies [5.i/5.ii]

Reviewer: ```suggestion
### Direct Runtime ROS Dependencies [5.i]/[5.ii]
```

> `rosidl_typesupport_cpp` does not have any runtime ROS 2 dependencies.

Reviewer: ```suggestion
`class_loader` does not have any runtime ROS 2 dependencies.
```

All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

> nightly_linux-/

Reviewer: Remove

Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)

> * [linux-arm64_release](https://ci.ros2.org/job/nightly_linux_release/lastBuild/testReport/class_loader/)

Reviewer: Rename this link because it's not arm64. Fix this one and the several below.
